streamer/src/nonblocking/quic.rs

```diff
@@ -960,37 +958,43 @@ async fn packet_batch_sender(
                 Ok(packet_receiver.recv().await)
             };
 
-            if let Ok(Ok(packet_accumulator)) = timeout_res {
+            if let Ok(Ok(mut packet_accumulator)) = timeout_res {
                 // Start the timeout from when the packet batch first becomes non-empty
                 if packet_batch.is_empty() {
                     batch_start_time = Instant::now();
                 }
 
-                unsafe {
-                    let new_len = packet_batch.len() + 1;
-                    packet_batch.set_len(new_len);
-                }
-
-                let i = packet_batch.len() - 1;
-                *packet_batch[i].meta_mut() = packet_accumulator.meta;
+                // 86% of transactions/packets come in one chunk. In that case,
+                // we can just move the chunk to the `Packet` and no copy is
+                // made.
+                // 14% of them come in multiple chunks. In that case, we copy
+                // them into one `Bytes` buffer. We make a copy once, with
+                // intention to not do it again.
+                // If that copy ever becomes a bottleneck, We could try to use
+                // some kind of chain or rope view over multiple `Bytes`
+                // buffers.
```

Comment:
doing any sort of chaining of different Bytes would break `transaction-view`; You can't get a `&[Signature] or `&[Pubkey]` if the data is not contiguous.

Reply:
Even if we are lucky enough that the signature would be entirely in the first chunk? I'm probably still not familiar enough with all the banking and scheduler code, but my impression from sigverify (I spend the vast majority of time on it and tests related to it) is that it relies on offsets, which in many cases might be in the first chunk.

Anyway, I'm fine with trimming the comment even more. I'm not expecting the "transaction rope" to be a relevant optimization anytime soon, even if it's possible.

Reply:
Trimmed down the comment.